test/e2e/clusterctl_upgrade.go

```diff
@@ -100,6 +100,9 @@ type ClusterctlUpgradeSpecInput struct {
 	// multiple infrastructure providers (ex: CAPD + in-memory) are installed on the cluster as clusterctl will not be
 	// able to identify the default.
 	InfrastructureProvider *string
+	// Allows to inject a function to be run after test namespace is created.
+	// If not specified, this is a no-op.
+	PostNamespaceCreated func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace string)
```

Comment:
Should we call this again in l.351? 

(same in self_hosted.go l.210)

Reply:
Nice catch, I forgot we have tests who are creating an additional management cluster.
Fixed (we are now calling the hook only for the namespace where we run the test)

For the management cluster itself, I think the correct way to go is what is suggested in https://github.com/kubernetes-sigs/cluster-api/issues/7613, but I would recommend keeping this discussion outside of this PR